### ISSUE TITLE
Collapsing is broken

### ISSUE THREAD

**Issue by addui** [adobe/brackets#12232]
Sometimes when I collapse and uncollapse code that has collapsible code within it (like a function with an if statement within it) everything gets all screwed up and the collapse arrows point the wrong direction and the collapsing just stops working all together.

**Comment by bradjohnwoods** [comment]
I get this all the time, very frustrating.

**Comment by ficristo** [comment]
I cannot reproduce with this small javascript snippet:
```javascript
function foo() {
    if (false) {
        return;
    }
}
```
Can you look at `Debug` --> `Show Develop Tools` when will happen again?
Will happen without extensions too?
If you have clear step to reproduce let us know (code, language, etc).

**Comment by addui** [comment]
If it was one specif example that would be one thing, but its everywhere all the time. Its screwed up more often than it works. I downgraded to Brackets 1.5 and it works perfectly again.

**Comment by bradjohnwoods** [comment]
I'm not sure how to reproduce exactly, but it happens all the time after maybe a couple of collapsing/ expanding actions. This is a screenshot from the devtools - 
![image](https://cloud.githubusercontent.com/assets/9333204/15061144/27762de2-1375-11e6-8a8b-1f6660c47d55.png)

**Comment by bradjohnwoods** [comment]
Is this issue easily fixable / will it be fixed soon? It's making the ide unusable. I went as far as to install an extension that was supposed to handle collapsing but brackets threw an error saying that extension was usable.

**Comment by addui** [comment]
Yes it makes this version of Brackets completely unusable. I downgraded to 1.5 and it works great, there are a few extensions I cant use but all the ones I want I can use so my suggestion would be to downgrade until this is fixed.

**Comment by addui** [comment]
So Brackets 1.7 came out today and I was all excited thinking I could start using the latest version of brackets again so I could get the latest plugins... NOPE you still have not fixed this issue, it's still happening. Your IDE is still completely unuseable because your collapsing is broken.

![brackes-issue](https://cloud.githubusercontent.com/assets/16481552/15978487/4110bf56-2f2c-11e6-9ecc-8f7424522c51.gif)

**Comment by ficristo** [comment]
The issue is clear, not how can we reproduce.
You could try to see if you can reproduce on [CodeMirror demo](http://codemirror.net/demo/folding.html) to see if an issue with it.
You could try to delete your state.json and brackets.json (remember to take a backup) and see if you can reproduce again.
/cc @thehogfather as the original author to see if he has better ideas.

**Comment by addui** [comment]
I can't reproduce it in the demo. It normally does not start happened until I've been using brackets for a 1/2 hour+, but with a long enough time using brackets its happens EVERY time. The log shows no errors. I'm just going to switch to another IDE.

**Comment by thehogfather** [comment]
@addui if you dont mind answering the following - would love to help out and sort this out when I get some time. Are you on a mac, windows or linux? And what version. Do you typically have more than one instance of brackets open when you encounter this bug? Approximately how many lines of code is in the file in the animated screen shot in you post.

P

**Comment by addui** [comment]
I am using Windows 10. I have experienced this error with Brackets 1.6 and 1.7. I may have more than one file open with Brackets but only ever 1 instance of Brackets open at once. That particular file has about 300 lines of code. Another thing I just thought of is that I always work off a network drive. The file never sits on my local machine but rather a server so that I can access it from multiple workstations, and yes this error occurs on multiple workstations (all windows 10).

**Comment by thehogfather** [comment]
@addui I have a tentative PR that addresses this issue #12521 - though I have no idea how long it would take the brackets team to review and possibly merge it. If you are comfortable with git and you've worked with brackets development before you can test the branch [https://github.com/thehogfather/brackets/tree/Code-folding-issue-12456](https://github.com/thehogfather/brackets/tree/Code-folding-issue-12456)

The issue has to do with the caching and syncing of the folded lines and that significantly changed across versions. If you are able to test this, that would be great - otherwise I might reconsider resurrecting the extension version of this so that I can make more frequent bug fixes.

**Comment by addui** [comment]
I'm just going to stick with Brackets 1.5 for now and would like to some kind of notification when this issue is fixed so that I can upgrade.

**Comment by designcise** [comment]
I was able to reproduce this by:

1. Folding some piece of code;
2. Selecting & hitting tab to indent the code;
3. Try collapsing or expanding the indented code, it should reproduce the problem.

I have witnessed this problem on other occasions as well, but like others have pointed out, haven't been able to reproduce it.

**Comment by alignsoft** [comment]
I'm seeing this issue as well with Brackets 1.7 on a Mac running OS X 10.11.5

Folding works fine for a while, but after some time the triangles stop responding, usually after unfolding a function to review, then the triangle indicates it's collapsed again even though the code is still visible, and after that I have to quit and relaunch Brackets to get folding working again.

Project is 40 files ±, ranging anywhere from a few dozen to 1,000 lines of code.

**Comment by thehogfather** [comment]
In the end the issue is quite easy to reproduce. All one has to do is the following:

1. Start with a file with say 40 lines with at least one foldable region.
2. Fold one foldable region
3. Anywhere above the foldable region, delete or add a line.
4. Try to unfold the folded region in 2.

Note that I have done the best I can to address this issue by submitting a PR #12521. It is a tiny change and should be easy to review. Happy to answer any questions about the context. I cannot do more than that. Someone with write permissions on the repo needs to review and decide whether or not to merge this. I am still investigating whether this bug goes all the way back to the CodeMirror api.

cc @abose @MarcelGerber

**Comment by abose** [comment]
tagging @swmitra

**Comment by derrickb** [comment]
Can confirm @thehogfather 's repro steps as well.  This issue has been driving me crazy.

I'm on Windows 10 x64, Brackets 1.7

**Comment by alignsoft** [comment]
Any chance this can get merged into the project? Is there a 1.7.1 update in the wings?

**Comment by nethip** [comment]
@alignsoft We could go ahead with the merge, but I am afraid we don't have 1.7.1 on the cards. Once the PR is merged, I would recommend you run Brackets from source. It's pretty easy to do that. Refer to the below link to do so.

https://github.com/adobe/brackets/wiki/How-to-Hack-on-Brackets

**Comment by addui** [comment]
So has the bug's cause been identified and fixed? Is it fixed in source? If not then there is no point in running from Source yet.

**Comment by ficristo** [comment]
Should be fixed in 1.8 by https://github.com/adobe/brackets/pull/12521
@addui you can try building brackets from source now that #12521 has been merged.

**Comment by addui** [comment]
Thank You, I will give it a shot when I get the chance
